### DISCUSSION TITLE
LayoutRes support in Aegisub

### DISCUSSION THREAD

**Discussion by arch1t3cht** [libass/libass#734] [Q&A] [answered]
**EDIT** for future reference: I was working under some misconceptions when writing these questions. If I recall correctly, I believed that LayoutRes effectively worked as a "PlayRes but better". But this is, of course, not the case: PlayRes and LayoutRes have distinct and orthogonal roles. LayoutRes does not replace PlayRes; what it effectively does is 1. say "This subtitle file was authored on a video with this resolution; please scale blur and the focal length for 3D rotations accordingly" and 2. opt into sane aspect ratio scaling. It is possible to change a file's PlayRes without changing the LayoutRes and vice-versa.

---

[TypesettingTools/libass](https://github.com/typesettingtools/libass/tree/meson-pr) was rebased to a version newer than 0.17.1 recently and pinterf/VSFilter also saw a new release supporting LayoutRes, so now all builds of TypesettingTools/Aegisub and arch1t3cht/Aegisub now support the LayoutRes headers in their subtitle renderers. Thus, it's probably time to also add proper support for setting them inside of Aegisub. I'd like to ask a couple of clarifying questions about this, but for future reference I'll outline the full current situation first:

There are four places in Aegisub that affect the PlayRes headers (apart from them just being passed on to the renderer like all other headers):
- When creating a new file with default settings
- When viewing/editing a file's headers via the "Script Properties" dialog
- In the "Resample Resolution" dialog
- When opening a video whose resolution does not match the script's resolution (in which case a dialog is opened asking the user on how to handle this mismatch. This is done in the `update_video_properties` function. [This function also silently sets the `YCbCr Matrix` header to the video's detected color matrix, and I'm not sure if that is the best way to handle things, but that's a discussion for another day])

Apart from that, the values are also *read* for the `subs.script_resolution` field in the Lua API and for computing coordinates to use in the visual typesetting tools.

Naturally, the question is how these usages should be updated to account for LayoutRes. One simple possibility would be to replace PlayRes with LayoutRes everywhere and automatically make the values of PlayRes match those of LayoutRes internally. Thus, my first big question is: Would this be correct? I know that the ASS File Format guide says to make PlayRes match LayoutRes for newly authored subtitles, but are there cases involving existing scripts where users would reasonably want LayoutRes to have a different value from PlayRes or be explicitly unset?

The second question is: How should mismatches between Play/Layout/Video resolution be handled now? For example:
- Should files with a PlayRes but no LayoutRes automatically be updated to set LayoutRes? (There's little precedent for Aegisub silently changing file properties, but I'm still leaning towards yes. If users really do not want to set LayoutRes I believe they can set it to 0 to explicitly opt out.)
- Should there still be a "Resolution Mismatch" dialog shown when opening a script whose LayoutRes differs from the video resolution? (I'm thinking not, since there afaik there now are no real problems with such a resolution mismatch. Also, showing such a dialog would be annoying in the case of anamorphic video where LayoutRes is set to the display aspect ratio rather than the storage aspect ratio.)
- Should there be a warning or error shown when opening a file whose PlayRes differs from the LayoutRes?
- There might exist old automation scripts that update PlayRes and are unaware of LayoutRes. Should that get any special handling? (I myself am not aware of any scripts like that so it's probably best to not worry about that for now. In theory a check for this could be implemented, after which the user could be asked whether or not this should be corrected.)

(I realise that some of those questions might be out of scope for this discussion since they're very specific to Aegisub, but I wanted to lay out all my thoughts here for the sake of completeness. Feel free to ignore some of the questions if necessary.)

This should be everything for now. As you can see, I have ideas for solutions for most of the questions, but I'd like to get a second opinion first to see if I missed anything or got something wrong.

**Comment by TheOneric** [comment, marked as answer]
To keep this thread (hopefully) easily understandable for any current or future onlookers, I’m going to repeat and explain some things you probably already know.

> I know that the ASS File Format guide says to make PlayRes match LayoutRes for newly authored subtitles, but are there cases involving existing scripts where users would reasonably want LayoutRes to have a different value from PlayRes or be explicitly unset?

First up: if the desired layout resolution can be inferred, I believe explicitly setting the headers is strongly preferable. As for `PlayRes ≠ LayoutRes`:

If `LayoutRes` is already set and differs from `PlayRes` the difference must be kept, although i see no reason why anyone would want to create *new* files like this.

In case of old files without `LayoutRes` headers, any standard ASS file will want its effective layout resolution matching the video’s storage size. Presumably for most releases the storage resolution and `PlayRes` will match. But there are at least two categories of releases to look out for:

1. Old, `LayoutRes`-less anamorphic files. They needed either explicit `\fscx` or `\fscy` overrides in all text segments to correct for anamorphic desqueezing.
   I’m guessing most such files will also use the storage size for `PlayRes`, but it is possible the native display size was used instead and in that case the discrepancy must be kept.  
   I can’t  spot any conclusive research into whether such files exist for sure, but this came up during #645., since apart from `PlayResX` this PR also fixed anamorphic scaling. Before _some_ features would look right on libass only if `PlayRes AR = DAR` and _some_ others only if `PlayRes AR = SAR`.

2. Poorly QCed, automated releases, reusing an identical script for several video resolution. I’d *hope* nowadays those set `LayoutRes` to make things actually work without user overrides, but many such files predate `LayoutRes`.  
    Afaik most of them will indeed at least have `PlayRes` matching the desired layout resolution, but I’m told there also exist half-baked resamples which now expect a layout resolution different from both `PlayRes` *and* the storage resolution of the video they’re bundled with, fun.  
    Those are hopefully the only case where determining the implicitly expected layout resolution isn’t straightforward.  
    However, I don’t think those files have much if any significance for Aegisub’s defaults and can probably be ignored in the following.

> Should files with a PlayRes but no LayoutRes automatically be updated to set LayoutRes? (There's little precedent for Aegisub silently changing file properties, but I'm still leaning towards yes. If users really do not want to set LayoutRes I believe they can set it to 0 to explicitly opt out.)

I think this generally makes sense, yep. I guess one tricky case might be if a user is porting subs to a higher res source (with the original potentially being an anamorphic DVD) and immediately opening subs with the new video.  
Perhaps, if `LayoutRes` is unset and `PlayRes` disagrees with both video storage and display res a dialogue could be displayed asking the user whether to initialise to `PlayRes`, storage res, display res or a custom resolution? *(Coincidentally this should also cover most but not all shovelware releases)*  
If they match, silently initialising seems fine to me.

> Should there still be a "Resolution Mismatch" dialog shown when opening a script whose LayoutRes differs from the video resolution? 

If the aspect ratio of `LayoutRes` matches the storage aspect ratio of the video (SAR), or potentially *(see case 1 above)* the display aspect ratio (DAR) of the video, I don’t think such a dialogue makes sense. It’s presumably a properly reused script and any fixups it might need for pixel-perfect alignment on the new resolution should be relatively minor *(... ignoring possibly needed colour primaries and gamut adjustments)*. But changing `LayoutRes` will entail major adjustments up to redoing all type transforms.

If however it matches neither, then yes I think displaying dialogue might still be a good idea. This likely (only?) happens when someone is porting subs from say an older 4:3 source to a new *(hopefully not just cropped)* widescreen release. It is unlikely subs whose, margins, linebreaks and types were authored for a different AR will be even close to good on a different AR source, so a major overhaul will be required regardless. Plus the mismatch of aspect ratios of `LayoutRes`and the new video’s DAR would lead to text being stretched, which is clearly undesirable.

> Should there be a warning or error shown when opening a file whose PlayRes differs from the LayoutRes?

If both are set explicitly *(and don’t trigger a previous case)*, I can’t imagine how changing any of them would improve things.

> Apart from that, the values are also read for the `subs.script_resolution` field in the Lua API and for computing coordinates to use in the visual typesetting tools.

`PlayRes` is still the correct coordinate system to position and scale most things in. `LayoutRes` only affects stuff which previously relied on the video’s storage resolution. If any scripts assumed they can use `PlayRes`for affected features it seems to me like they were already broken before.  
Overriding whatever fields expose the storage resolution with `LayoutRes` might make sense though.

> There might exist old automation scripts that update PlayRes and are unaware of LayoutRes. Should that get any special handling?

I’m afraid I can’t say and have no idea what options there are to detect and hook into this.

**Comment by astiob** [comment]
Sorry; I started writing a response two days ago but lost it to a crash and got a bit discouraged.

---

As a recap, the following things are affected by `LayoutRes` instead of or in addition to `PlayRes`. You may want to update the Resolution Resampler to account for this and to put this in documentation:

* `\blur` is simply scaled from `LayoutRes`. In current libass, it is scaled independently along each axis, accurately supporting anamorphic releases. In current xy-VSFilter/XySubFilter and older libass (and possibly current MPC-HC), it is scaled from `LayoutResY` only (along both axes).

* Text (i. e. fonts but not drawings) is stretched based on the `LayoutRes` aspect ratio (in addition to `\fscx`). `Fontsize` or `\fs` is scaled from `PlayResY` along both axes and laid out onto the `LayoutRes` rectangle, then transformed together with this rectangle to display resolution. If the display aspect is wider than `LayoutRes`, then the text is widened (stretched horizontally); if narrower, then it is narrowed.

* `\frx` & `\fry` 3D transforms use a fixed camera distance of 312.5 px (20000÷64) scaled from `LayoutResY`, and the resulting perspective transform is also likely affected by the `LayoutRes` aspect ratio. Being angles rather than distances, these tags cannot be simply rescaled for a different `LayoutRes`. Thus, if a file contains any of these transforms, you generally want to keep the `LayoutRes` forever unchanged or the transforms will break.

* `\be`, not being a distance value, likewise cannot be simply value-scaled for a different resolution. Due to this, DirectVobSub renders it directly at video resolution, making it conceptually a layout-resolution feature. However, even as the other tags have been fixed in modern renderers, `\be` is so hard to scale that it’s _still_ not scaled at all in libass or MPC-HC, and while XySubFilter does attempt to scale it and uses `LayoutResY` for it, it doesn’t actually achieve an accurate effect. I have an old libass branch that does this properly (at the expense of a slight quality loss/extra blur), but it needs to be brought up to date and may need to be sped up via assembler. If this ever happens, it will also be scaled from `LayoutRes`; after all, we don’t want to introduce a _third_ set of resolution headers just for `\be`. If Aegisub gets something like a compliance/compatibility checker, it would be good to warn if a file contains `\be` when its `LayoutRes` differs from the associated video’s storage size.

Thus: if a file contains `\frx` or `\fry`, you want to keep the `LayoutRes` constant regardless of other changes unless the user wants to redo the transforms by hand or using appropriate Lua scripts (until such scripts’ functionality is integrated into Aegisub—although in any case rescaling rotations can produce ugly output tags); otherwise, Aegisub is free to resample automatically and can adjust `LayoutRes` to match `PlayRes` for better user intuition, although if the file contains `\be`, manual intervention is still required to fix that. (For `\be` with very small argument values, Aegisub could perhaps automatically replace it with `\blur`, as the difference is nonzero—and variable across renderers—but not huge. For large values though, there’s no guarantee that the effect is even reproducible in ASS at the new resolution at all.)

> are there cases involving existing scripts where users would reasonably want LayoutRes to have a different value from PlayRes […]?

Absolutely. That’s why the headers exist in such form in the first place. The idea is that you pick up an existing release and tag it with its video size in `LayoutRes` before doing anything else to the script. Then you’re free to slap that script on different videos, resample it to a different `PlayRes` in Aegisub, etc. and it should keep looking the same (at least as long as there’s no `\be`).

> replace PlayRes with LayoutRes everywhere and automatically make the values of PlayRes match those of LayoutRes internally

You could indeed do this, i. e. change `PlayRes` to match `LayoutRes` rather than the other way around. This would work. However, if their file contains any nonrescalable layout-dependent tags, this would lock the user into a resolution they may not be a fan of.

On a rare occasion, an advanced user might want to intentionally supersample `PlayRes` to get access to more granular `Margin` values, which are required to be integers, or other values where ASS supports fractional values but Aegisub only exposes integer UI. Of course, by doing so while using `LayoutRes`-dependent features in the same file, they also run the risk of breaking the rendering even harder in at least some of the older players that don’t know about `LayoutRes`.

> Should files with a PlayRes but no LayoutRes automatically be updated to set LayoutRes?

IMO, not until you have a video. And even then, the best thing would be to offer the user to set `LayoutRes` to the video size but at the same time allow them to specify a custom resolution instead.

Aegisub has no way of knowing whether the video is really the same video the script was originally authored for. If the user’s intention is to reuse an older script for their own re-release, they’re quite likely to be opening their new video rather than the original one (unless they’re actually opening the subs and the video together from a Matroska file directly within Aegisub). Even if they are opening an original file, it may have been part of a multiresolution release where the subs were only actually accurate for one of the resolutions, or a Crunchyroll release where the subs were targeting a fixed resolution different from any of the video sizes.

But in any case, when a video is loaded into Aegisub, having _some_ `LayoutRes`, even if not the one the user truly wants, ensures that the current rendering they’re seeing on the preview screen will stay unchanged even if they swap in a different video.

Regardless of how you initially set `LayoutRes`, I do think you want to allow the user to edit it manually to correct for mistakes.

> are there cases […] where users would reasonably want LayoutRes to […] be explicitly unset?

Off the top of my head, no.

There’s one interesting case with libass-targeting files: `ScaledBorderAndShadow` is unset or even explicitly set to `no`, then in player setups that don’t call `ass_set_storage_size` (either because the player was never updated to set it or because the user has manually turned it off to improve compatibility with files like this), current libass actually does scale borders and shadows as if the header were set to `yes`—but only if `LayoutRes` is also unset (otherwise `LayoutRes` assumes the role of the storage size call). Thus, if the file’s video size differs from its ASS `PlayRes` (and it contains any LayoutRes-affected ASS features), the behaviour actually differs between an unset `LayoutRes` and any particular `LayoutRes` value. But I do think such files are fixable by changing the header to `yes` and then adding the `LayoutRes`. (On that note, libass also defaults an unset `ScaledBorderAndShadow` to `yes` for some files, namely FFmpeg conversions, regardless of `LayoutRes`.)

> There might exist old automation scripts that update PlayRes and are unaware of LayoutRes. Should that get any special handling?

It’s anyone’s guess, of course, but I’d imagine that such scripts are attempting some kind of script rescaling for a different video. In that case, they probably do indeed want the `LayoutRes` to be updated, too (and they probably treat the original `PlayRes` as if it were also the original `LayoutRes`, although they may also let the user input the layout resolution in a dialog box).

**Comment by arch1t3cht** [comment]
It's been quite a while, but thanks for your answers. I now finished my first draft of fully supporting LayoutRes in Aegisub https://github.com/TypesettingTools/Aegisub/pull/594 (though the majority of changes in that PR are actually related to YCbCr Matrix instead). In brief, I opted to just ask the user in almost all cases (at least by default; all prompts can be configured) but provide sane defaults. Once the Aegisub manual gets some love, I will try to write a detailed guide explaining all of the subtleties of PlayRes, LayoutRes, and YCbCr Matrix.

**Reply by TheOneric** [reply]
Great work, from the description this sounds like a substantial improvement and tricky UX indeed ^^

Just one remark from skimming the changes, it looks like the option for (not) being asked about LayoutRes changes are either never, if no LayoutRes defined yet, if LayoutRes is defined but does not exactly match video resolution, or both of the before.  

If I got this right, it may be worth considering to differentiate the "mismatch" case further into "not matching exactly" (like currently) and "mismatching AR (but exact values for individual dimension are allowed to differ)" since keeping the old LayoutRes is expected/encouraged and unproblematic if the AR matches *(except for `\be` or when targeting old renderers)*, but if the AR differs likely introduces more issues then it helps.  
Being asked each time about this when opening a project based on a reused script from a lower res release might lead to the prompt just being disabled fully and then later missing an actual AR mismatch *(if the difference is small enough to not be immediately obvious)*

**Reply by arch1t3cht** [reply]
Thanks for pointing this out, I added a "When aspect ratio changes" option now and made that the default, and made the prompt change depending on whether the aspect ratio changed or stayed the same.
I'm not yet sure what extra actions should be offered to the user when the aspect ratio changes (it's more or less the same question as how to handle LayoutRes when resampling), but for now the user is at least notified and can decide themselves.
